### PR TITLE
Dholladay00/sap ramp

### PR DESCRIPTION
**Pull request by dholladay00**
## PR Summary
This PR will introduce a new modifier that enables ramping an EOS in a way that is compatible with ramp implementations in downstream host codes. 

This PR is still in a very draft state, WIP PR put in place for comments since there are a few possible paths of pursuit. This can provide a location for discussion.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [ ] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.

### CODE REVIEW

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:122]: should we cover more cases?

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:154]: specific overloads might be necessary

**Review by Yurlungur — COMMENTED** [review]
Overall this seems reasonable and basically dirt simple, which I'm a fan of.

**Comment by jhp-lanl** [comment]
I'll give this a thorough review tomorrow

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:122]: this is copy/paste code from shifted eos. I will work finishing updating the code when I get a chance.

File: singularity-eos/eos/singularity_eos.cpp
Comment (dholladay00) [inline singularity-eos/eos/singularity_eos.cpp:56]: This is how I'm currently handling the multiple ways to initialize a ramp that covers standard initialization and initialization from p-alpha model parameters. If this looks ok, I'll do this for all of the other Eos's. Would like some eyes maybe @Yurlungur and @jhp-lanl

Comment (Yurlungur) [reply singularity-eos/eos/singularity_eos.cpp:56]: This seems fine to me. The names are a little vague, though. What is `vals` vs `vals_i`?

Comment (Yurlungur) [reply singularity-eos/eos/singularity_eos.cpp:56]: I think letting them be non-const is the cleaner and clearer design

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:148]: I _believe_ most implementations of a ramp utilize the slope of the ramp itself in the bulk modulus calculation

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:200]: Won't the vector calls just access the scalar versions where the modifier is already applied?

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:159]: We'll need P-T initialization for this to be used in EAP. Since the ramp is bilinear, I'd think it would be pretty easy to write a function to invert it so that you can get rho(P) instead of P(rho).

The tricky part will be figuring out if you're on the ramp or not without knowledge of `r1`. Maybe you can call `FillEos` assuming no ramp and then compare the output density of that to the value you get out of the ramp and see which one is greater. If the EOS density is greater than the ramp then you're on the ramp.

So maybe something like
```c++
if (output & thermalqs:density) {
  assert(!(output & thermalqs::pressure));  // Need pressure if density isn't specified
  Real temp_in = temp; 
  Real energy_in = energy;
  Real press_in = press;
  t_.Filleos(rho, temp_in, energy_in, press_in, cv, bmod, lambda);
  const Real rho_ramp = get_ramp_density(press);
  if (rho > rho_ramp) {
    rho = rho_ramp;
  }
}
```
There may be more I'm missing though like how to modify the bulk modulus...

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:137]: for my own edification, is there a particular reason to use curly bracket initialization here rather than just the assignment operator?

**Review by jhp-lanl — CHANGES_REQUESTED** [review]

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:159]: > We'll need P-T initialization for this to be used in EAP. Since the ramp is bilinear, I'd think it would be pretty easy to write a function to invert it so that you can get rho(P) instead of P(rho).

@jdolence and I actually got a request from Ristra recently for a general API for this for all singularity-eos EOS's. I think it's a great opportunity for the base class. We can have a default functionality that uses a numerical root find and overload as needed.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:137]: Yes, like I said (even though I'm inconsistent with it), it is a good habit to have. Though for POD types it's not needed.

Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:137]: I don't think in this context it matters much... Despite being around @dholladay00 for a long time now, I still haven't picked up this habit for scalars. :P

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:159]: > > We'll need P-T initialization for this to be used in EAP. Since the ramp is bilinear, I'd think it would be pretty easy to write a function to invert it so that you can get rho(P) instead of P(rho).
> 
> @jdolence and I actually got a request from Ristra recently for a general API for this for all singularity-eos EOS's. I think it's a great opportunity for the base class. We can have a default functionality that uses a numerical root find and overload as needed.

You mean putting a general version of `FillEos` in the base class? At the very least we definitely need to standardize its behavior across all of the classes, but yes I agree that a general implementation in the base class would be a good idea. I'll create an issue for this.

But regardless, I would still think we need a special `FillEos` here since the ramp complicates things.

Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:159]: > You mean putting a general version of FillEos in the base class? 

I meant providing a `rho(P, T)` and/or `rho(P, e)` function in the base class. That way it's available for `FillEos` to use for individual EOS's. But yeah I agree we need to standardize behaviour

Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:159]: > > You mean putting a general version of FillEos in the base class?
> 
> I meant providing a `rho(P, T)` and/or `rho(P, e)` function in the base class. That way it's available for `FillEos` to use for individual EOS's. But yeah I agree we need to standardize behaviour

Ah okay, yes those are probably good ideas

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:148]: how do I tell when to use the ramp slope vs the eos evaluated one?

Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:148]: You probably need to do the ramp pressure vs EOS pressure test to figure out if you're on the ramp or not

Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:148]: To add to this, most ramps assume that the ramp defines a roughly isentropic path for the material which makes the slope of the ramp a direct input (i.e. no thermodynamic relations needed) to the isentropic bulk modulus calculation.

Whether this is physical or not is another question, but if we want to match the current implementations I would just use the slope of the ramp.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: @jhp-lanl could you take a look at the most recent changes and see if they address your concerns?

Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: @jhp-lanl could you take a look at the most recent changes and see if they address your concerns?

File: singularity-eos/eos/singularity_eos.cpp
Comment (dholladay00) [reply singularity-eos/eos/singularity_eos.cpp]: @Yurlungur this has now been addressed.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:195]: Starting a new code comment to keep things clean.

I think the logic here is backwards. The ramp is used to decrease the EOS density so you only want to use it when it's less than the EOS density.

**Review by jhp-lanl — CHANGES_REQUESTED** [review]
Just one issue in the comparison of the ramp and EOS densities. After that's fixed I approve.

File: utils/scripts/format.sh
Comment (jhp-lanl) [inline utils/scripts/format.sh:4]: Are we changing the copyright again?

**Review by jhp-lanl — CHANGES_REQUESTED** [review]
Sorry one more thing I noticed: there are no tests for the ramp itself and only that the modifier doesn't do anything with default settings (a good test!).

I would propose three tests:
1. Verify B_S and P on the first leg of the ramp
2. Verify B_S and P on the second leg of the ramp
3. Verify that B_S and P are the same as the EOS for a density beyond the ramp

File: utils/scripts/format.sh
Comment (Yurlungur) [reply utils/scripts/format.sh:4]: oh oops. No this was on me. I copy-pasted this script out of Phoebus.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:41]: This is a conversation for a later time---but with the new PTE solver, do we *need* `ValuesAtReferenceState`? If not, perhaps we should deprecate it and remove it.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:209]:
```suggestion
    printf("Ramp parameters: r0=%.14e\na=%.14e\nb=%.14e\nc=%.14e\n", r0_, a_, b_, c_);
```

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Overall looks good. Some nitpicks below. I also agree with @jhp-lanl that this needs tests of the ramp physics.

Also @jhp-lanl how do you feel about helping us fill out the ramp physics in the docs?

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:174]: Oops. Already an issue. #135.

File: singularity-eos/eos/singularity_eos.hpp
Comment (Yurlungur) [reply singularity-eos/eos/singularity_eos.hpp:31]: oops I asked for this and forgot. Resolved.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:182]: Why not just call `PressureFromDensityInternalEnergy`? Why dereference `this`?

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/modifiers/ramps_eos.hpp:188]: Won't pressure be overwritten here?

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:91]: There are asserts in the ctor, but perhaps we should disallow r0 = 0.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:97]: there are asserts in the ctor for the ramp parameters.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:188]: notice that I pass in the `ramp_out` which ensures pressure is not an output value.

Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:188]: ah ok I see.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:91]: We should disallow r0=0 if we're dividing by it. But even small r0 could cause problems.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (Yurlungur) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:97]: You check for exactly zero. We should have division by zero guards because a small number divided by a small number can still give the wrong answer.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:43]: It should be greater than or equal to one, but I realize one of the tests currently relies on it being less than 1.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:91]: I believe the tests will cause `r0` to be fairly small since they are ramping an ideal gas. The basic idea is that $\rho_0 = \rho(P_0, T_\mathrm{ref}) / \alpha_0$. For an ideal gas, the density will be on order 1e-03.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:41]: We at least need the density at reference state in order for us to initialize the parameters from p-alpha parameters.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: done

File: test/test_eos_unit.cpp
Comment (jhp-lanl) [inline test/test_eos_unit.cpp:97]: This isn't a hold on the MR, but one thing I've been trying to do is use the Catch2 `INFO` function since `REQUIRE(isClose(...))` will just return `false` without any more helpful information. The `INFO` macro is a print statement that will execute only when the test fails so you can get some more information about why it's failing. This has _massively_ helped in my debugging of the Gruneisen EOS because just knowing the two values can help me determine what's going crazy.

The other thing to consider is using `CHECK` rather than `REQUIRE` for a big block like this. `CHECK` fails the test still if the condition is false, but it won't end the test at the first failure. This can be helpful for debugging when you can see a pattern in how multiple values are off.

File: test/test_eos_unit.cpp
Comment (jhp-lanl) [inline test/test_eos_unit.cpp:313]: Can you break this up into multiple `THEN` statements? I think you just need to transform your comments into `THEN` statements so that you have five tests instead of one long one.

**Review by jhp-lanl — CHANGES_REQUESTED** [review]
Small changes

File: test/test_eos_unit.cpp
Comment (jhp-lanl) [inline test/test_eos_unit.cpp:101]: It looks like a bunch of these tests have nothing to do with the fast logarithms even if they're grouped there. I'm going to create an issue to break these out

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:92]: I'd say that number is not small relative to machine epsilon. I am wary of judging relative orders of magnitude in a way that prevents users from running the code without a special build. I will try to add @Yurlungur's smart divides thing, but I do not want to further restrict densities besides r0 needs to be a positive and non-zero floating point value.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp:104]: I'll try to implement smart/robust divide, but I still prefer to use the ternary where possible. *I* had formatted the ternary in a readable way, but I'm not allowed to format. And I don't want to restrict r0 beyond the > 0 constraint.

**Comment by jonahm-LANL** [comment]
If you have a way you want to format, you can use
 // clang-format disable
On a piece of code.

Sent with BlackBerry Work
(www.blackberry.com)


From: Daniel Holladay ***@***.******@***.***>>
Date: Friday, Jul 01, 2022, 11:14 AM
To: lanl/singularity-eos ***@***.******@***.***>>
Cc: Subscribed ***@***.******@***.***>>
Subject: [EXTERNAL] Re: [lanl/singularity-eos] Dholladay00/sap ramp (PR #101)


@dholladay00 commented on this pull request.

________________________________

In singularity-eos/eos/modifiers/ramps_eos.hpp<https://urldefense.com/v3/__https://github.com/lanl/singularity-eos/pull/101*discussion_r912128188__;Iw!!Bt8fGhp8LhKGRg!EGzueB5w418lcM_PRvqsfAKthTD4veOegN7RxrvbBp61DRNSR8HtzW1OouRyR1_67qVCOt_PmuOExDzQeLh4sdi9$>:

> +  Real get_ramp_pressure(Real rho) const {
+    const Real p_ramp{rho < r0_     ? 0.0
+                      : rho < rmid_ ? a_ * (rho / r0_ - 1.0)
+                                    : b_ * (rho / r0_ - c_)};
+    return p_ramp;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real get_ramp_density(Real P) const {
+    const Real rho_ramp{P < Pmid_ ? r0_ * (P / a_ + 1.0) : r0_ * (P / b_ + c_)};
+    return rho_ramp;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real get_ramp_dpdrho(Real rho) const {
+    const Real dpdr{rho < r0_ ? 0.0 : rho < rmid_ ? a_ / r0_ : b_ / r0_};


I'll try to implement smart/robust divide, but I still prefer to use the ternary where possible. I had formatted the ternary in a readable way, but I'm not allowed to format. And I don't want to restrict r0 beyond the > 0 constraint.

File: test/test_eos_unit.cpp
Comment (dholladay00) [reply test/test_eos_unit.cpp]: done.

File: test/test_eos_unit.cpp
Comment (dholladay00) [reply test/test_eos_unit.cpp]: done.

**Review by jhp-lanl — APPROVED** [review]
Looks good to me. @Yurlungur has @dholladay00 satisfied your comments?

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: done

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: now using robust divide.

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: now using robust divide

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: now using robust divide

File: singularity-eos/eos/modifiers/ramps_eos.hpp
Comment (dholladay00) [reply singularity-eos/eos/modifiers/ramps_eos.hpp]: I have written up the math for converting from p-alpha to sap ramp.

**Review by Yurlungur — APPROVED** [review]
Assuming all tests pass on Darwin, yep this looks good to me.

**Comment by Yurlungur** [comment]
Thanks for all your hard work on this @dholladay00 can you submit a PR on Darwin to make sure tests pass there?

**Comment by dholladay00** [comment]
gpu tests failing, but not in places I changed. I'll take a look next week, DO NOT MERGE before hand. I think it could be issues with doing emplacement of the ramp params in the builder test and the value category of the indexer bracket operator in the pte test.

**Comment by dholladay00** [comment]
@Yurlungur @jhp-lanl gitlab ci fixes should work, I say once the latest ci finishes its good to merge.
Addenda:
@Yurlungur I added device markings to a couple spiner member functions to suppress warnings, may be worth checking out before clicking the button. 
@mauneyc-LANL maybe double-check my minor cmake changes are reasonable.

**Comment by Yurlungur** [comment]
@dholladay00 is there a Spiner MR? Or do you mean SpinerEOS?

File: singularity-eos/eos/eos.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos.hpp:537]: Making these `PORTABLE_INLINE_FUNCTION` won't do any harm. It's worth noting though that when these functions are called in a parallel context, it's undefined behavior.

File: test/test_eos_unit.cpp
Comment (Yurlungur) [inline test/test_eos_unit.cpp:70]: Not a problem, but worth noting that this will only work on CPU context.

**Review by Yurlungur — APPROVED** [review]
LGTM.

**Comment by Yurlungur** [comment]
Happy to merge upon gitlab success and approval by @mauneyc-LANL .

**Comment by dholladay00** [comment]
@Yurlungur I just remembered that @jhp-lanl had requested a name change `SAPRamp` -> `BilinearRamp` I'll start working on that.

**Comment by dholladay00** [comment]
> @Yurlungur I just remembered that @jhp-lanl had requested a name change `SAPRamp` -> `BilinearRamp` I'll start working on that.

This change has been completed, CI running now.

**Pull request merged**
